src/main/java/net/fabricmc/loom/task/RemapJarTask.java

```diff
@@ -27,276 +27,255 @@
-import net.fabricmc.loom.configuration.providers.mappings.MappingsProviderImpl;
-import net.fabricmc.loom.util.Constants;
-import net.fabricmc.loom.util.TinyRemapperHelper;
-import net.fabricmc.loom.util.ZipReprocessorUtil;
+import net.fabricmc.loom.extension.MixinExtension;
+import net.fabricmc.loom.task.service.JarManifestService;
+import net.fabricmc.loom.task.service.TinyRemapperService;
 import net.fabricmc.loom.util.ZipUtils;
-import net.fabricmc.stitch.util.Pair;
+import net.fabricmc.tinyremapper.InputTag;
+import net.fabricmc.tinyremapper.OutputConsumerPath;
 import net.fabricmc.tinyremapper.TinyRemapper;
-import net.fabricmc.tinyremapper.TinyUtils;
-import net.fabricmc.tinyremapper.extension.mixin.MixinExtension;
 
-public class RemapJarTask extends Jar {
+public abstract class RemapJarTask extends AbstractRemapJarTask {
 	private static final String MANIFEST_PATH = "META-INF/MANIFEST.MF";
 
-	private final RegularFileProperty input;
-	private final Property<Boolean> addNestedDependencies;
-	private final Property<Boolean> addDefaultNestedDependencies;
-	private final Property<Boolean> remapAccessWidener;
-	private final List<Action<TinyRemapper.Builder>> remapOptions = new ArrayList<>();
-	public JarRemapper jarRemapper;
-	private FileCollection classpath;
-	private final Set<Object> nestedPaths = new LinkedHashSet<>();
+	@InputFiles
+	public abstract ConfigurableFileCollection getNestedJars();
+
+	@Input
+	public abstract Property<Boolean> getAddNestedDependencies();
 
+	@Inject
 	public RemapJarTask() {
 		super();
-		LoomGradleExtension extension = LoomGradleExtension.get(getProject());
-		input = getProject().getObjects().fileProperty();
-		addNestedDependencies = getProject().getObjects().property(Boolean.class)
-				.convention(false);
-		addDefaultNestedDependencies = getProject().getObjects().property(Boolean.class)
-				.convention(true);
-		remapAccessWidener = getProject().getObjects().property(Boolean.class)
-				.convention(false);
-
-		if (!extension.getMixin().getUseLegacyMixinAp().get()) {
-			remapOptions.add(b -> b.extension(new MixinExtension()));
-		}
+
+		getClasspath().plus(getProject().getConfigurations().getByName(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME));
+		getAddNestedDependencies().convention(true).finalizeValueOnRead();
 	}
 
 	@TaskAction
-	public void doTask() throws Throwable {
-		boolean singleRemap = false;
-
-		if (jarRemapper == null) {
-			singleRemap = true;
-			jarRemapper = new JarRemapper();
-		}
+	public void run() {
+		final LoomGradleExtension extension = LoomGradleExtension.get(getProject());
 
-		scheduleRemap(singleRemap || LoomGradleExtension.get(getProject()).isRootProject());
+		submitWork(RemapAction.class, params -> {
+			if (getAddNestedDependencies().get()) {
+				params.getNestedJars().plus(getNestedJars());
+			}
 
-		if (singleRemap) {
-			jarRemapper.remap();
-		}
-	}
+			params.getRemapAccessWidener().set(extension.getAccessWidenerPath().isPresent());
```

Comment:
IMO this should rather read it from the fabric.mod.json than check the extension. This approach has some issues with multiproject jars where
- project A has an AW
- project B has a shadowed copy of project A's `namedElements`, so the AW would still be `named`
- project B declares A's AW in its fabric.mod.json but it's not remapped because project B doesn't declare it
(one example of these project setups is Adorn, where I have to unnecessarily apply the AW to "project B" and create a second MC jar when only "project A" needs it)

Reply:
I have made it always remap the AW if one is present in the jar, I dont see a usecase to remap the classes and not the AW.